image-fip.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Ahmad Fatoum <a.fatoum@pengutronix.de>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <confuse.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "genimage.h"
+
+static int fip_generate(struct image *image)
+{
+	struct partition *part;
+	char *args = strdup("");
+	const char *extraargs = cfg_getstr(image->imagesec, "extraargs");
+	int ret;
+
+	list_for_each_entry(part, &image->partitions, list) {
+		struct image *child = image_get(part->image);
+		char *oldargs;
+
+		oldargs = args;
+		xasprintf(&args, "%s--%s '%s' ", args, part->name, imageoutfile(child));
+		free(oldargs);
+	}
+
+	ret = systemp(image, "%s create %s%s '%s'", get_opt("fiptool"),
+		      args, extraargs, imageoutfile(image));
+
+	free(args);
+
+	return ret;
+}
+
+static const char *tos_fw[] = { "tos-fw", "tos-fw-extra1", "tos-fw-extra2" };
+
+static int fip_setup(struct image *image, cfg_t *cfg)
+{
+	unsigned int num_tos_fw;
+
+	num_tos_fw = cfg_size(cfg, "tos-fw");
+	if (num_tos_fw > ARRAY_SIZE(tos_fw)) {
```

Comment:
`parse()` is called before `setup()` so this check happens too late.
So move the check to `parse()` and remove `setup()` (it's optional).